autogen/middleware/base.py

```diff
@@ -0,0 +1,226 @@
+import asyncio
+import functools
+import inspect
+from functools import wraps
+import threading
+from typing import Any, Callable, List, Optional, Protocol
+
+from typing_extensions import TypeVar
+
+__all__ = ["Hookable", "hookable"]
+
+F = TypeVar("F", bound=Callable[..., Any])
+H = TypeVar("H", bound=Callable[..., Any])
+T = TypeVar("T")
+
+
+class Middleware(Protocol):
+    def call(self, *args: Any, **kwargs: Any) -> Any:
+        ...  # pragma: no cover
+
+
+class Hookable(Protocol):
+    """Protocol for hookable functions and methods."""
+
+    @property
+    def _origin(self) -> Callable[..., Any]:
+        ...  # pragma: no cover
+
+    @property
+    def _chained_call(self) -> Callable[..., Any]:
+        ...  # pragma: no cover
+
+    @property
+    def _is_async(self) -> bool:
+        ...  # pragma: no cover
+
+    @property
+    def _is_method(self) -> bool:
+        ...  # pragma: no cover
+
+    @property
+    def _middlewares(self) -> List[Middleware]:
+        ...  # pragma: no cover
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        ...  # pragma: no cover
+
+
+def hookable(f: F) -> Hookable:
```

Comment:
Since we are moving away from hooks to middleware. Perhaps we can use a different name. How about `core`, `core_method`, or `core_function`.